### ISSUE TITLE
improve support for surface data in nilearn?

### ISSUE THREAD

**Issue by jeromedockes** [nilearn/nilearn#2171]
It seems that there is growing interest for surface data in nilearn and
nistats, for example:

-   <https://github.com/nistats/nistats/issues/401>
-   <https://github.com/nilearn/nilearn/issues/2132>
-   <https://github.com/nilearn/nilearn/issues/2128>
-   <https://github.com/nilearn/nilearn/issues/2423>

and nilearn already has functions for surface plots.

To move forward in this direction, we may want to add a proper way to
represent, mask and unmask, read and write surface data &#x2013; similar to
`Nifti1Image` and `NiftiMasker`. For now surface data is passed around as
separate arrays for the meshes and the data.

I guess nibabel's [`GiftiImage`](https://nipy.org/nibabel/reference/nibabel.gifti.html#nibabel.gifti.gifti.GiftiImage) is meant for that, although I am not sure how
it works and haven't found any examples. Maybe @juhuntenberg has used it
before?

@kchawla-pi @bthirion @GaelVaroquaux @juhuntenburg

**Comment by dangom** [comment]
Is it an idea to also support Freesurfer surfaces?

**Comment by bthirion** [comment]
Yes, because they are supported by nibabel.

**Comment by sahahn** [comment]
I think there is certainly benefit in a more universal Surface class similar to Nifti1Image, but it definitely isn't straightforward. I.e., nibabel has the GiftiImage, for working with Gifti files, but for Freesufer support just has loading functions vs. an actual class ( https://nipy.org/nibabel/reference/nibabel.freesurfer.html). Then there is cifti case... which seems like it would need to be handled differently.

It seems like there are some functions where passing data separate from the associated geometry mesh is okay, i.e., in cases where you only are performing an operation on the data itself, like extracting ROIs or processing time-series w/ signal.clean. On the other hand, there are certain functionalities where knowing meta-data around what 'space' the surface is in would be required, e.g., if you wanted to allow resampling between different spaces (as an analog to the automatic volumetric re-sampling), though I think this would be a rabbit-hole to try and implement.

Likewise, because of the standard in outside software (Freesurfer and HCP) of keeping the data separate from the meshes, maybe it isn't a great idea to try and side-skirt that. 

I'm definitely curious on other's thoughts.

**Comment by GaelVaroquaux** [comment]
The problem of the babel tower of surface file formats is one that we hit
when designing the surface plotting code.

We created a few helper functions to load the data when we did this. 
https://github.com/nilearn/nilearn/blob/master/nilearn/surface/surface.py#L558

**Comment by jeromedockes** [comment]
@emdupre and I had a discussion about this today and we think that we should try
to handle surface images similarly to volume images, i.e. manipulate Gifti
images that contain both the mesh(es) and the data array(s).

Just as Nifti images have a data array, a shape and an affine, gifti objects
would have one or more data arrays and a matching number of meshes (vertex
coordinates and triangles), and a GiftiMasker would transform that into a single
1d (or 2d for timeseries) array. This avoids manipulating value arrays that are
not anchored to any position in 3d space. The masker is the interface between
images used for io and plotting and the 2d arrays of signals used for analysis.

If we assume that users take care of this themselves and provide nilearn with
data arrays, a lot of functionality is available through lower-level functions
or directly scikit-learn.

One possible difficulty is that unlike affines, meshes are relatively large, but
as they are not modified and in many analyses only one mesh is involved memory
mapping can help.

Regarding IO and file formats, nibabel would handle that and give us a uniform
interface through GiftiImage in the same way that it does with NiftiImage, and
if we want to work to better support different file formats and use cases, this
would ideally be done through contributions to nibabel.

We also think that we should leave cifti aside for now, and focus first on good
support for gifti-like data. Finally, we don't think we should attempt to
provide any kind of resampling of surface data (any time soon). As a side note,
volume to surface conversion could also be provided as some kind of masker or
transformer to be included in pipelines.

Disclaimer: I have never worked with surface data and am completely naive on
that topic. This is not the case for @emdupre.

We would also like to hear @effigies, @bthirion and @GaelVaroquaux opinions on
this.

**Comment by bthirion** [comment]
I like your approach: the model is consistent when the spatial context is saved along with the "texture" data. Just beware that currently these things are most often saved on disk separately (but this is not an issue for the proposed design IIUC).

**Comment by jeromedockes** [comment]
> I like your approach: the model is consistent when the spatial context is saved along with the "texture" data. Just beware that currently these things are most often saved on disk separately (but this is not an issue for the proposed design IIUC).

Indeed I don't think this is an issue. And I guess they are likely to
keep being stored separately in the future, to avoid storing several
copies of meshes.

**Comment by sahahn** [comment]
> I like your approach: the model is consistent when the spatial context is saved along with the "texture" data. Just beware that currently these things are most often saved on disk separately (but this is not an issue for the proposed design IIUC).
> Indeed I don't think this is an issue. And I guess they are likely to keep being stored separately in the future, to avoid storing several copies of meshes.

I think the GiftiImage interface is a bit clunky right now when it comes to storing both texture and mesh arrays together, just since most people would be loading in initially two separate GiftiImages. I.e. you can load them both separately then add what they call the GiftiDataArrays from one to the other (but then it isn't clear how you would go about merging all of the other info stores, i.e. meta data, absence or presence of a labeltable, etc...). Also when it comes time to save a transformed image, I imagine people would want to save just the texture.

I guess also to be discrete, if you don't see adding in support for surface based resampling anytime soon, then where would the mesh even be used? For plotting? Spatial smoothing might be reasonably easy to implement? I guess my point just being, we would want to make sure that there is actually a reason to have the mesh explicitly associated with the texture, especially since most softwares handle them separately.

One more consideration for why storing textures with meshes might be a bit funny is that a texture can work with a number of meshes, i.e. associated sphere coordinates, pial surface, flattened representation, etc... Such that if the GiftiImage were being used in a masking context and let's say smoothing was requested, you might want the associated sphere, but then later on in the pipeline when it comes time to plot, you might want to use a different mesh. 

Not sure if I articulated my hesitation well...

**Comment by jeromedockes** [comment]
Those are some good points. The main reason for keeping texture and mesh
together is so that the user is not responsible of keeping track of
which data point corresponds to which node in which mesh (e.g. in which
order arrays corresponding to two hemispheres have been concatenated).
Nifti images are still more convenient to work with than plain data
arrays even if you don't need any resampling.
But indeed the fact that meshes are large and that we don't perform
resampling for surface data are important differences between volume and
surface, and may make this approach impractical for surfaces.

> I think the GiftiImage interface is a bit clunky right now when it comes to storing both texture and mesh arrays together, just since most people would be loading in initially two separate GiftiImages. I.e. you can load them both separately then add what they call the GiftiDataArrays from one to the other (but then it isn't clear how you would go about merging all of the other info stores, i.e. meta data, absence or presence of a labeltable, etc...). Also when it comes time to save a transformed image, I imagine people would want to save just the texture.

some of this can be addressed in nibabel or by wrapping some of
nibabel's functionality. e.g. to_filename could have a store_mesh kwarg

>
> I guess also to be discrete, if you don't see adding in support for surface based resampling anytime soon, then where would the mesh even be used? For plotting? Spatial smoothing might be reasonably easy to implement? I guess my point just being, we would want to make sure that there is actually a reason to have the mesh explicitly associated with the texture, especially since most softwares handle them separately.
>

I guess it is true that for most of the processing you don't need the
mesh, but the same could be said about Nifti images

> One more consideration for why storing textures with meshes might be a bit funny is that a texture can work with a number of meshes, i.e. associated sphere coordinates, pial surface, flattened representation, etc... Such that if the GiftiImage were being used in a masking context and let's say smoothing was requested, you might want the associated sphere, but then later on in the pipeline when it comes time to plot, you might want to use a different mesh. 
>

that's true, but you could just create a new image

> Not sure if I articulated my hesitation well... 

I think you did, and indeed there are some valid arguments for mostly
ignoring the meshes. I would like contributors who have experience
working with surface data to weight in.

At any rate how to represent surface data should be decided before
adding code geared towards surfaces. especially since if we only use the
texture, adding support for surface data is mostly exposing a low-level
api that deals with numpy arrays.

**Comment by sahahn** [comment]
> Those are some good points. The main reason for keeping texture and mesh
> together is so that the user is not responsible of keeping track of
> which data point corresponds to which node in which mesh (e.g. in which
> order arrays corresponding to two hemispheres have been concatenated).
> Nifti images are still more convenient to work with than plain data
> arrays even if you don't need any resampling.
> But indeed the fact that meshes are large and that we don't perform
> resampling for surface data are important differences between volume and
> surface, and may make this approach impractical for surfaces.

I agree that it would be nice to be abstract away keeping track of the correspondence between data point and mesh location - though in some cases it might be a bit unavoidable, e.g., in my previous example I raised where at first you want to use the sphere for sampling, but then later on want to use a different mesh... the user would need to either know how the data points correspond (e.g., left hemisphere concat'ed with right, so they can later on provide that same order of meshes) - or would have had to pass both meshes when creating the original surface. 

One way a lot of existing software seems to handle to issue of correspondence seems to be just through either meta-data (i.e., its easier to just mark a file as in 'fsaverage5' space then save all of the associated data about fsaverage5 with that subject, but at the same time it relies on the software still having access to all that data in some way), or the just saving tons of files (e.g., FreeSurfer). Maybe there is a way in the Gifti/Surf wrapper class we could support by default some of the most common spaces- but also provide an interface for passing in user passed meshes (i.e., for working with surface data still in a subjects native space, there is no way to support that apriori). 
Another thing which might be helpful is just making sure the Gifti/Surf wrapper class we use is somewhat flexible, i.e., like the GiftiImage class, you can have a GiftiImage with just texture data or both texture data and mesh data, and depending on the function it is passed too, maybe the GiftiImage doesn't require a mesh- so that function will still work even if the GiftiImage isn't saved with a mesh.

> some of this can be addressed in nibabel or by wrapping some of
> nibabel's functionality. e.g. to_filename could have a store_mesh kwarg

Yeah that's true. I think a pretty light wrapper around GiftiImage could go a long way in terms of usability.

> I guess it is true that for most of the processing you don't need the
> mesh, but the same could be said about Nifti images

True. I guess what I mean though is I just think it would be useful to define ahead of time a rough list of functions that it would eventually be nice to have surface support for. That way we can make sure the surface representation we settle on is flexible enough to cover all cases (obviously easier said then done). 

> that's true, but you could just create a new image

Yeah, and honestly the plotting case is really the only one I can think of where it would show up in the first place.

> At any rate how to represent surface data should be decided before
> adding code geared towards surfaces. especially since if we only use the
> texture, adding support for surface data is mostly exposing a low-level
> api that deals with numpy arrays.

I very much agree! I'll keep thinking more about it, and looking forward to seeing what other people think too. My initial leanings though are that a high level wrapper over some existing nibabel functionality would be really helpful, i.e., there's no reason why a user should have to convert their freesurfer surfaces to gifti before being able to use them, or end up having to write a loop that wraps all of their texture files with a single standard mesh

**Comment by SylvainTakerkart** [comment]
Hi,

If I'm not mistaken, a GiftiMasker or SurfaceMasker class that would act as the NiftiMasker one would require a functionality that would allow the re-interpolation of textures between different meshes (e.g from fsaverage to fsaverage5), right? In the volume, it's always possible! But with surfaces, this is not always possible (e.g you can interpolate between fsaverage and fsaverage5; but you cannot interpolate between fsaverage and a native mesh / the sphere, at least without having access to a bunch of other info, i.e some surface2surface transfo). This will remain a fundamental difference between surface and volume handling, right?

Anybody has given any thoughts about this? Or is it handled by the design proposed above by @jeromedockes and @emdupre ? (I did not get all the details)

Also, we'd given all this some thoughts with @juhuntenburg a while ago...

Cheers,

**Comment by GaelVaroquaux** [comment]
> if you don't see adding in support for surface based resampling anytime soon, then where would the mesh even be used? For plotting?

Plotting is not really optional.

I don't know much about surface, but something that is critical to the
design here is that it should be easy to use for a newcomer.

All my experience with surface softwares is that they are difficult to
use, because they require a certain organisation of file or data, and
this organisation is implicit.

We need to maintain together the basic elements that make a data usable
and understandable.

**Comment by jeromedockes** [comment]
> This will remain a fundamental difference between surface and volume handling, right?

Yes, you are correct -- there will be no resampling AFAICT. Of course this
makes a surface  much less powerful than the volume maskers, but surface
masker objects could still be useful despite this limitation.

> Also, we'd given all this some thoughts with @juhuntenburg a while ago...

Is there by any chance a trace of this in an issue somewhere? or do you
remember reaching any conclusion?

**Comment by bthirion** [comment]
I think that considering that only fsaverage meshes are supported is OK, at least as a first step.

**Comment by sahahn** [comment]
> I think that considering that only fsaverage meshes are supported is OK, at least as a first step.

I like this idea - what about allowing a user to just specify what space the surface is in with a string, i.e., similar to the params for loading the fsaverage surface mesh, like mesh = fsaverage5. That said... we could likely even infer common spaces based on the number of vertex.

**Comment by dangom** [comment]
I would love to see surface data support in nilearn as well.  I'm not an expert in surface analysis, so I will write my opinion as a beginner, but I've been playing with Freesurfer surfaces on and off for the past year, such that I have a general idea of what the file structure look like, and I am  familiar with the nilearn volumetric API. I must also say I know nothing about giftis, so please apologize if I say something wrong and please correct me.

> I don't know much about surface, but something that is critical to the
> design here is that it should be easy to use for a newcomer.
> 
> All my experience with surface softwares is that they are difficult to
> use, because they require a certain organisation of file or data, and
> this organisation is implicit.

I totally agree with this statement. I think, however, that every newcomer must first understand that proper surface analysis is nontrivial and it is more complex that volumetric analysis, if anything simple because data is acquired in volume space, such that surfaces are (almost by definition) derived from volumes.

Next, the surface lingo should be made clear for newcomers and consistent throughout nilearn, so a glossary would come in handy when we think about what a SurfaceMasker should look like. From my beginner understanding of Freesurfer and nilearn, I would say that (again, please correct me if I'm wrong):

* Glossary of surface terms

**mesh** - _refers to the vertices and triangles that define a surface_
Each vertex has an index and an (x, y, z) coordinate. Each triangle has a (v1, v2, v3) coordinate, where v1, v2 and v3 correspond to vertex indices. The x y z coordinates only make sense with respect to the coordinate system of the volume from which the surfaces were generated.
In Freesurfer, white and pial surfaces have the same number of vertices, and there is a topological correspondence between them. Very loosely speaking, vertices form pseudo-columns across the cortical ribbon. Interpolation between surfaces may seem like a reasonable idea, but it's actually quite hard to do properly. Simply "averaging" pial and white surfaces may give us something close to a "mid-layer", but that is not what Freesurfer would do because this average may not respect smoothness and topological constraints (such as self-intesection, for example). Respecting these constraints is important for the idea of inflation and registration: we won't be able to register a surface to another if one is a torus and the other a sphere.

**texture** - _refers to values associated to a mesh_
Each texture contains a list of pairs <vertex, value> (not necessarily implemented as such). A texture only makes sense with respect to the surface from which it was generated. In Freesurfer, a texture would be called an "overlay".

In the tentative SurfLabelsMasker PR, a mesh is called a Nisurf-mesh, and a texture is called Nisurf-data. I think the "2D Nisurf-data" refers to an array of 1D Nisurf-data, correct?

```
we may want to add a proper way to
represent, mask and unmask, read and write surface data – similar to
Nifti1Image and NiftiMasker. For now surface data is passed around as
separate arrays for the meshes and the data.
```

* What should a SurfaceMasker look like?

TLDR: In my opinion a SurfaceMasker takes a mesh and a reference volume and allows projecting volumetric data onto the surface. It ideally also allows data to be backprojected into a volume using `inverse_transform` (and saving data into a texture object, such as a Gifti?).

I would like to argue for a SurfaceMasker to require a mesh and to keep a reference to the original volume from which the mesh was generated. In Freesurfer, this information is stored in the file structure -- surfaces such as `surf/lh.white` being in the space of `mri/orig.mgz`. In nilearn, that information would be passed as an argument. A possible derived class FreeSurfaceMasker could possibly infer that from the FreeSurfer file structure.
Providing a SurfaceMasker without a reference to a volume does not help newcomers, for they would need to deal with a spatial reference outside of the masker class, which requires knowledge of the organisation of file or data, making nilearn yet another software difficult to use. 

A `SurfaceMasker.fit_transform(volume)` would return the equivalent of a texture object, or the underlying array by projecting the volume onto the surface properly account for registering the fitting volume to the volume in which the surface mesh is.

Here I don't know what an "inverse" transform should optimally look like. In NiftiMasker, we get our processed data back into 3D. Should a `SurfaceMasker` simply return a gifti, or backproject the processed texture into the volume (kind of like what surf2vol) does? I think the latter would be more convenient for newcomers.

For masking surfaces, what exactly do we want to achieve? If the goal is simply to constrain the computation for a given region, then should the masking be applied to the volume being projected onto the surface first, or do we simply tag vertices outside the mask as "masked", similar to numpy's masked arrays?

**Comment by jeromedockes** [comment]
> TLDR: In my opinion a SurfaceMasker takes a mesh and a reference volume and allows projecting volumetric data onto the surface. It ideally also allows data to be backprojected into a volume using `inverse_transform` (and saving data into a texture object, such as a Gifti?).
>

I think that would be useful to have, but we probably want something
that handles surface data provided by the user as well. So we might end
up with bot a GiftiMasker (transform surface data to 1d array) and a
VolToSurfMasker (basically a pipeline of volume to surface projection
then GiftiMasker), or something similar.

> I would like to argue for a SurfaceMasker to require a mesh and to keep a reference to the original volume from which the mesh was generated. In Freesurfer, this information is stored in the file structure -- surfaces such as `surf/lh.white` being in the space of `mri/orig.mgz`. In nilearn, that information would be passed as an argument. A possible derived class FreeSurfaceMasker could possibly infer that from the FreeSurfer file structure.
> Providing a SurfaceMasker without a reference to a volume does not help newcomers, for they would need to deal with a spatial reference outside of the masker class, which requires knowledge of the organisation of file or data, making nilearn yet another software difficult to use. 

yes, that is exactly what we want to avoid

>
> A `SurfaceMasker.fit_transform(volume)` would return the equivalent of a texture object, or the underlying array by projecting the volume onto the surface properly account for registering the fitting volume to the volume in which the surface mesh is.
>

we don't handle registration, but it could do resampling if both images
are in the same space but sampled on different grids.

> Here I don't know what an "inverse" transform should optimally look like. In NiftiMasker, we get our processed data back into 3D. Should a `SurfaceMasker` simply return a gifti, or backproject the processed texture into the volume (kind of like what surf2vol) does? I think the latter would be more convenient for newcomers.
>
> For masking surfaces, what exactly do we want to achieve? If the goal is simply to constrain the computation for a given region, then should the masking be applied to the volume being projected onto the surface first, or do we simply tag vertices outside the mask as "masked", similar to numpy's masked arrays?
>

probably just concatenate the textures into a 1d array (e.g. if there
are two for both hemisphere), with the possibility to exclude certain
nodes. the main goal is to have an object that remembers to which mesh
node each value in the 1d output corresponds. and then of course there
are the variants such as mapsmasker, labelsmasker etc.

**Comment by juhuntenburg** [comment]
Hi, I have been a bit silent but been following this and am looking forward to the discussion later today! (as mentioned here https://github.com/nilearn/nilearn/issues/2423#issuecomment-631288161)

**Comment by GaelVaroquaux** [comment]
A tentative proposal from the discussion today:

# Rational

Nilearn wants to keep together enough information about surface data to enable plotting and spatial processing (eg smoothing on the surface), but would like to avoid as much as possible creating a new data structure.

A challenge is that common practice (eg in freesurfer, or connectome pipelines) relies on storing geometry and textures separately.

The goal is to have an interface that allows easily move this data around, for instance as input and output of the future SurfaceMasker, or to plot.

# Proposal

## The Surface

Main data structure: a `texture, geometry`, stored by default as namedtuple
```
Surface = namedtuple('surface', ['texture', 'geometry'])
surf = Surface(texture=GiftiImage(...), geometry=GiftiImage(...))
```
As input, we would also accept variants, such as a tuple instead of a named tuple, and filenames to gifti or mgz file instead of in-memory files. As output, we would always return the above.

## The Geometry

Geometry must itself contain `points` and `triangles`, which are numpy arrays. They can be accessed via `.points` and `.triangles` attributes of the geometry input, or with the right nibabel incantation:
```
points, triangles = GiftiImage().agg_data(('triangle', 'pointset'))
```
We will need a bit more work to make this work on freesurfer files. Nibabel will try to evolve to facilitate accessing this information.

# Road forward

1. Validate and amend the above proposal
2. Implement or evolve loader functions in nilearn to make it easy to cast the above input data structure to the 3 corresponding numpy arrays, and back to the canonical structure
3. Use this in our existing surface code (beware to keep backward compat):
    - plotting
    - projection from vol to surf
4. Use this in the surface masker

The last points above can be implemented in parallel.

**Comment by bthirion** [comment]
texture is a numpy array or necessarily a `GiftiImage` ?

**Comment by jeromedockes** [comment]
I think numpy array should be accepted

**Comment by juhuntenburg** [comment]
Thanks for putting it together @GaelVaroquaux . A few notes:
> Main data structure: a texture, geometry, stored by default as namedtuple

I am not sure if a namedtuple had a order, if it does I think the order should be geometry, texture

> Implement or evolve loader functions in nilearn to make it easy to cast the above input data structure to the 3 corresponding numpy arrays, and back to the canonical structure

I assume that we need to implement new functions to not break backward compatibility? Also, we have to agree on a nomenclature. Here we are using "geometry" and "texture", which I am happy with, but which is at odds with what exsits in Nilearn currently ("mesh" and "data"). So an easy way to move forward would be to implement new functions that are load_geometry and load_texture and start deprecating the old ones? Or is that not the right approach?

**Comment by jeromedockes** [comment]
> I am not sure if a namedtuple had a order, if it does I think the order should be geometry, texture

It does

**Comment by bthirion** [comment]
In terms of naming I prefer `mesh` to `geometry`.

**Comment by juhuntenburg** [comment]
I will make a first attempt to adjust the loading functions to this proposal now

**Comment by effigies** [comment]
I'm not entirely thrilled with textures... Is a vector or time series a texture? A collection of textures?

Also, just to be clear, aren't textures per face, not per vertex?

**Comment by sahahn** [comment]
I like the design!

A few clarification points:
- So we are going forward with the GiftiImage as an internal representation? Then further, all output would be as the namedtuple, with the two stored GiftiImages, right?

- We want to accept filenames as inputs still right? I.e, so in summary, you can pass 
for texture:
A GiftiImage (w/ internal checks to make sure it contains data), a numpy array (either 1D or 2D time-series) or a filepath to be loaded as a gifti.
For the geometry: 
A GiftiImage (w/ internal checks to make sure it has the triangle and the pointset), a tuple of numpy arrays, formatted correctly (i.e, what mesh in plotting accepts now), or again a file path.

Is that a fair summary, or am I missing some cases?

- About adding FreeSurfer support, it seems like this could either be done better in nibabel (i.e., implemented in the way that @chrisgorgo talked about in the discussion, where it would have its own object w/ complementary functions to the GiftiImage) - or, in the short term we could allow FreeSurfer input by just converting it internally to GiftiImage's (though this would force them to be loaded into memory)

**Comment by sahahn** [comment]
> I'm not entirely thrilled with textures... Is a vector or time series a texture? A collection of textures?
> 
> Also, just to be clear, aren't textures per face, not per vertex?

I thought in FreeSurfer at least, textures were one value for vertex, and represented some combination of all faces that include that vertex. I could be wrong though

**Comment by juhuntenburg** [comment]
> I'm not entirely thrilled with textures... Is a vector or time series a texture? A collection of textures?
> 
> Also, just to be clear, aren't textures per face, not per vertex?

For now I am sticking with "mesh" and "data"  
But happy to adapt if there are other preferences

**Comment by juhuntenburg** [comment]
@sahahn 

No, I think idea is that a mesh is a namedtuple of two numpy arrays (coordinates and faces) and a surfaces is a named tuple of a said mesh and data (which I think is simply a numpy array)

We will continue to accept freesurfer file formats and gifti as inputs, as well as lists, tuples and namedtupled

**Comment by sahahn** [comment]
> @sahahn
> 
> No, I think idea is that a mesh is a namedtuple of two numpy arrays (coordinates and faces) and a surfaces is a named tuple of a said mesh and data (which I think is simply a numpy array)
> 
> We will continue to accept freesurfer file formats and gifti as inputs, as well as lists, tuples and namedtupled

I'm okay with that too, but it seems to contradict what @GaelVaroquaux posted above? Maybe I read his post wrong

**Comment by jeromedockes** [comment]
> For now I am sticking with "mesh" and "data"  
> But happy to adapt if there are other preferences

+1 for "mesh" and "data"

**Comment by juhuntenburg** [comment]
> > @sahahn
> > No, I think idea is that a mesh is a namedtuple of two numpy arrays (coordinates and faces) and a surfaces is a named tuple of a said mesh and data (which I think is simply a numpy array)
> > We will continue to accept freesurfer file formats and gifti as inputs, as well as lists, tuples and namedtupled
> 
> I'm okay with that too, but it seems to contradict what @GaelVaroquaux posted above? Maybe I read his post wrong

Yeah actually the way you wrote the example @GaelVaroquaux might be a bit misleading. Inserting a GiftiImage is one way of constructing a surface, but it can also be just a filename or numpy arrays. At least this is how I understand it

**Comment by sahahn** [comment]
> > > @sahahn
> > > No, I think idea is that a mesh is a namedtuple of two numpy arrays (coordinates and faces) and a surfaces is a named tuple of a said mesh and data (which I think is simply a numpy array)
> > > We will continue to accept freesurfer file formats and gifti as inputs, as well as lists, tuples and namedtupled
> > 
> > 
> > I'm okay with that too, but it seems to contradict what @GaelVaroquaux posted above? Maybe I read his post wrong
> 
> Yeah actually the way you wrote the example @GaelVaroquaux might be a bit misleading. Inserting a GiftiImage is one way of constructing a surface, but it can also be just a filename or numpy arrays. At least this is how I understand it

My impression was/is we are okay accepting a wide variety of input, i.e., numpy arrays, named tuples, files, nibabel objects - but we want one internal representation, that is also the standard format returned to the user.

I think I do remember that during out discussion you or someone else floated the idea of using the GiftiImage as that internal representation, but I don't remember if there was consensus on an alternative. 

It seems like the options are either GiftiImage wrapped in named tuple, or numpy arrays wrapped in named tuples.

**Comment by SylvainTakerkart** [comment]
> For now I am sticking with "mesh" and "data" But happy to adapt if there are other preferences
> +1 for "mesh" and "data"

+1 also for "mesh" and "data"
(I like "data", it's less technical than texture, and it nicely carries the message that what we want to do with surfaces in nilearn is to analyse this "data")

**Comment by juhuntenburg** [comment]
@sahahn Yes, I was quite in favour of using GiftiImage as an internal representation but was convinced that it is too clunky for what we need. Also, @effigies was in favour of a simpler Surface object, to be pushed also in Nibabel. So what I understand is that we agreed on a namedtuple of numpy arrays

**Comment by sahahn** [comment]
> @sahahn Yes, I was quite in favour of using GiftiImage as an internal representation but was convinced that it is too clunky for what we need. Also, @effigies was in favour of a simpler Surface object, to be pushed also in Nibabel. So what I understand is that we agreed on a namedtuple of numpy arrays

Great. Looking forward to helping with this

**Comment by JulienLefevreMars** [comment]
+1 for data:

Maybe it could allow to use less trivial format than numpy.arrays 
For instance, if you consider a vector field, it can be seen as a Npoints x 3 array (coordinates of the vector at each point), but it is probably better to see it as a "set" of 2 coordinates attached to a local basis of the tangent space. 
Probably the same could be used for diffusion fibers.

**Comment by juhuntenburg** [comment]
@JulienLefevreMars not sure I understand, what format would you need for that?

**Comment by jeromedockes** [comment]
> (I like "data", it's less technical than texture, and it nicely carries the message that what we want to do with surfaces in nilearn is to analyse this "data")

also it is similar to `get_data` and `get_fdata` used for Nifti1Images

**Comment by JulienLefevreMars** [comment]
@juhuntenburg . Well, instead of Npoints x 3 numpy arrays, it could be something like
vectorField = namedtuple('vectorField', ['localCoordinates', 'localBasis'])
'localCoordinates' corresponds to Npoint x 2 array
'localBasis' is Npoint x 3 x 2 array, (2 stands for the dimension of the tangent plane, 3 for the coordinates of the basis vectors

Mathematically it is more satisfying but I don't know if it is better from the point of view of data representation.

**Comment by juhuntenburg** [comment]
Interesting, I think it depends on how common this is, whether or not it makes sense to implement it as a special case, or just be ok with a simple numpy array. I assume that the data file does not contain the information that what it contains is a vector field, so it would require a flag or argument to the loading function. 

I would be curious to hear others' opinions on this.

**Comment by JulienLefevreMars** [comment]
@juhuntenburg To be even more precise, the local basis information would me even more relevant in the 'mesh'/'geometry' part. But by writing this, I realize that it makes things really more complicated for the user (not supposed to be a mathematician).

**Comment by bthirion** [comment]
Hi @JulienLefevreMars 
I think that the intent here was to support very basic operations on such data: accessing, doing spatial filtering and various statistical computations.
I'm not sure whether this is relevant for the vectorField structure you describe above. there are probably (many) other operations you want to do ?

**Comment by JulienLefevreMars** [comment]
Hi @bthirion 
Spatial filtering is clearly relevant, if you think for instance to smooth a vector field given by the principal curvature. But if I am not wrong it is not equivalent to smooth a 3D vector field and to smooth the coordinates in the tangent bundle (the proper way to smooth).
It was just an example of data that could not be directly seen as "simple" arrays. There might be others (provided by tractography/connectivity ?).

**Comment by bthirion** [comment]
We won't introduce sphisticated smoothing schemes (I mean: beyond scalar field smoothing along the surface). I think that the kind of operations that you would like to do with vector fields on the surface won't be implemented in Nilearn. MindBoggle would be a better fit to support advanced manipulations ?

**Comment by JulienLefevreMars** [comment]
Thanks @bthirion for the explanation and the suggestion.

**Comment by juhuntenburg** [comment]
I changed the mesh loading function to return a namedtuple in #2474 

Next I would like to change the surface plotting functions to take a tuple / namedtuple as input. 
Currently these functions take two inputs (mesh and optional data). I assume I have to write new functions to keep backward compatibility? Or is there a solution I am overlooking?

If I write new functions, is there any convention / advice on what to call them, as they will essentially do the same as the current plot_surf etc ?

**Comment by GaelVaroquaux** [comment]
A slightly modified version of the proposal based on the feedback, changing a few things, and hopefully making everything more clear and explicit.

# Rationale

Nilearn wants to keep together enough information about surface data to enable plotting and spatial processing (eg smoothing on the surface), but would like to avoid as much as possible creating a new data structure.

A challenge is that common practice (eg in freesurfer, or connectome pipelines) relies on storing geometry and textures separately.

The goal is to have an interface that allows easily move this data around, for instance as input and output of the future SurfaceMasker, or to plot.

# Proposal

## The Surface

Main data structure: a `texture, geometry`, stored by default as namedtuple
```
Surface = namedtuple('surface', ['mesh', 'data'])
surf = Surface(mesh=GiftiImage(...), data=GiftiImage(...))
```
Internal checks validate that the GiftiImages contain mesh/data information.

Outputs of functions are always the above.

For inputs, we accept the following variants: 
* a tuple instead of a named tuple
* filenames to gifti or mgz file instead of in-memory files.
* MGH objects
* data: as an `np.array(shape=(t, p))`
* mesh: variants described below


## The mesh

The mesh must itself contain `points` and `triangles`, which are numpy arrays of. They can be accessed via `.points` and `.triangles` attributes of the geometry input (and as a result we accept named_tuples), or with the right nibabel incantation:
```
points, triangles = GiftiImage().agg_data(('triangle', 'pointset'))
```
We will need a bit more work to make this work on freesurfer files. Nibabel will try to evolve to facilitate accessing this information.

# Road forward

1. Validate and amend the above proposal
2. Implement or evolve loader functions in nilearn to make it easy to cast the above input data structure to the 3 corresponding numpy arrays, and back to the canonical structure
3. Use this in our existing surface code (beware to keep backward compat):
    - plotting
    - projection from vol to surf
4. Use this in the surface masker

The last points above can be implemented in parallel.

**Comment by juhuntenburg** [comment]
Thanks @GaelVaroquaux 

> Internal checks validate that the GiftiImages contain mesh/data information.

This is confusing to me. Do the entries to mesh have to be GiftiImages after all? I thought they would just be numpy arrays to avoid the complexity of the GiftImages object

**Comment by GaelVaroquaux** [comment]
> This is confusing to me. Do the entries to mesh have to be GiftiImages after all?

I have modified the text to be more clear. Tell me if it's clear.

The default object should be GiftiImages, in my opinion. As input because this is what people will easily get from disk, as output so that they can easily write them to disk.

**Comment by bthirion** [comment]
LGTM.

**Comment by juhuntenburg** [comment]
I just don't understand how you describe a mesh on one hand as a namedtuple of numpy arrays, but then when inputting it to a surface it needs to be a GiftiImage. Is mesh a namedtuple or a GiftiImage?

**Comment by juhuntenburg** [comment]
And consequently, should load_surf_mesh return a namedtuple or a GiftiImage?

**Comment by GaelVaroquaux** [comment]
>Is mesh a namedtuple or a GiftiImage?

The suggestion is that it can be either as an input. However, the output should always be a GiftiImage, to facilitate the saving of the data.

**Comment by juhuntenburg** [comment]
Because we decide that gifti is going to be the default for saving? Somehow I thought the whole idea of namedtuples was to stay agnostic to file formats

**Comment by GaelVaroquaux** [comment]
I see two reason to favor using gifti instead of plain numpy arrays:
1. Easier to carrying around meta information, which we may decide to add in the future
2. Easier to save.

For the second point, advanced users can always change the format. The point being that given a named tuple, beginners might struggle figuring out what to do about it. Also, I would hope that it's easy to save a GiftiImage in MGH with nibabel. But maybe I'm naive. @effigies , any comments?

**Comment by effigies** [comment]
If you want to use GiftiImage internally, writing a routine to write to a FreeSurfer geometry file should be simple, though I don't think the routine currently exists. The [`freesurfer.write_geometry`](https://nipy.org/nibabel/reference/nibabel.freesurfer.html#write-geometry) function elaborates what *can* be put in the file.

Anyway, I think it's going to be pretty rare that you actually want to save geometry, since you're not generating that with nilearn.

**Comment by jeromedockes** [comment]
> Next I would like to change the surface plotting functions to take a tuple / namedtuple as input. 
> Currently these functions take two inputs (mesh and optional data). I assume I have to write new functions to keep backward compatibility? Or is there a solution I am overlooking?
>
> If I write new functions, is there any convention / advice on what to call them, as they will essentially do the same as the current plot_surf etc ?
>

that's a difficult one :/

- changing the function themselves to make them magically accept both
  kinds of inputs would be a mess

- creating functions with different names in the same namespace would be
  problematic too, because as you point out the best names are already
  taken.

- I would suggest placing the new functions in a new module, and changing
  the imports in `plotting/__init__.py` at the end of the deprecation cycle

- There should also be an easy way to start using the new functions
  right away, maybe with a special import similar to those in
  `sklearn.experimental`( see https://scikit-learn.org/stable/modules/classes.html#module-sklearn.experimental),
  or a function call. however importing from the full path to the new
  module is also not super complicated. 

- The old module could continue existing for a longer deprecation cycle
  so that users who haven't had time to change all their function calls
  can keep their code working just by changing imports for some time


WDYT? curious to hear everyone's opinion

**Comment by juhuntenburg** [comment]
> If you want to use GiftiImage internally, writing a routine to write to a FreeSurfer geometry file should be simple,

Sure, but I would be a simple, if not simpler, to write a routine to save a namedtuple to a FreeSurfer geometry file.

**Comment by jeromedockes** [comment]
> Anyway, I think it's going to be pretty rare that you actually want to save geometry, since you're not generating that with nilearn.

it could still be useful, for example if a function receives the mesh as
an in-memory object and doesn't know where it is on disk, but needs to
write a complete image with data + mesh
we would probably need it for testing the loading functions as well

**Comment by juhuntenburg** [comment]
I am not principally against the suggestion to use GiftiImages, I just did not understand this as the conclusion of our discussion and just wanted to understand better. I think the advantage as @GaelVaroquaux said could be possibly associating metadata (although I never really used metadata with surfaces) and that saving to gifti will be easier. So if our general idea is to default to gifti that makes sense

**Comment by bthirion** [comment]
> Anyway, I think it's going to be pretty rare that you actually want to save geometry, since you're not generating that with nilearn.

Well, you may want to create a middle mesh by averaging white and pial coordinates, and save it.

**Comment by SylvainTakerkart** [comment]
> > Anyway, I think it's going to be pretty rare that you actually want to save geometry, since you're not generating that with nilearn.

I have the same feeling...

> 
> Well, you may want to create a middle mesh by averaging white and pial coordinates, and save it.

but that's starting to be some "mesh processing" operation, and I don't think nilearn's goal is to do "mesh processing", no?

**Comment by bthirion** [comment]
Well, I've already done it, and found it useful for display. This is very basic...

**Comment by SylvainTakerkart** [comment]
sure, sure, I don't question the usefulness of this nor its simplicity, just the relevance of it being part of nilearn... that would be an outlier functionality in nilearn, no? and therefore, maybe, that should not be a good enough reason to orient the design of some other piece of nilearn?

but maybe there are other reasons to save a mesh?
(I don't see any obvious one, but...)

**Comment by GaelVaroquaux** [comment]
I agree that saving meshes is not really going to happen much. Maybe it would make sense if we want to save full reports to disk, but that's something that a user would not do.

To me, however, it's as much about nilearn and about wider good practice of the ecosystem. I would find for never have volumetric data carried around as numpy arrays but rather as NiftiImage, as numpy arrays loose the meaning of the data.

In a wider context of defining standards for moving meshes around Python code, it seems to that a GiftiImage is a better choice, but I'm open to being corrected. I'd really like to hear @effigies on that.

**Comment by bthirion** [comment]
AFAIK GiftiImage does not bring more features than arrays, but I may not know it well enough.

**Comment by sahahn** [comment]
> AFAIK GiftiImage does not bring more features than arrays, but I may not know it well enough.

This seems to get at something we've been circling around for a while now - that no one seems fully satisfied with the GiftiImage's, but at the same time we don't want to as Gael keeps pointing out pass around bare numpy arrays. 

I.e., it would be great if there was already a well-implemented and accepted nibabel surface equivalent to the NiftiImage, but in lieu of one, we have to either use the GiftiImage (which has its problems), move towards generating a new standard (outside of the scope of nibabel), or pass around pure numpy arrays  (bad ecosystem practice).

Is this a fair summary ?

**Comment by bthirion** [comment]
> - I would suggest placing the new functions in a new module, and changing the imports in `plotting/__init__.py` at the end of the deprecation cycle

Sounds like the best solution to me. Thx !

**Comment by Remi-Gau** [comment]
Will close as this seems kind of obsolete.

Mentioned in the meta-issue #4027 in case we need more context.
